lib/resty/upstream/healthcheck.lua

```diff
@@ -435,6 +474,49 @@ local function get_lock(ctx)
     return true
 end
 
+local function set_all_down_flag(ctx, ppeers, bpeers)
+    local dict = ctx.dict
+    local u = ctx.upstream
+    local u_key = gen_upstream_key(u,"all_down")
+
+    local all_down, err = dict:get(u_key)
+    if not all_down then
+        errlog("Failed to get all down flag for upstream " .. u .. ". ", err)
+        return
+    end
+    
+    if all_down == 0 then
+        local result = true
+        
+        for i = 1, #ppeers do 
+            result = result and ppeers[i].down
+        end
+
+        for j = 1, #bpeers do
+            result = result and bpeers[j].down
+        end
```

Comment:
Can we maintain a counter for the number of up'd nodes so that we do not have to scan all the peers' state upon every checking operation? The current way looks too much expensive to me. An `O(1)` operation now becomes an `O(n)` operation.

Reply:
But the status of any peer can change after any check. Here we really need to go through all of them. 
The only one improvement I can see here is to stop when we found the first 'up' peer.
Added break condition into both loops.
I think it doesn't change the theoretical complexity, but of course in normal situation ( when all peers are up) it will be O(1).